File: device_agent/tools/device_tools.py
"""
Device Tools - Get/set device state, list devices.
"""

from typing import Any, Dict

from ..core.tool import Tool, ToolResult
from ..core.runtime import AgentRuntime
from ..types import Device, DeviceState


class DeviceStateGetTool(Tool):
    """Tool to get device state."""

    name = "device_state_get"
    description = "Get the current state of a device"
    input_schema = {
        "type": "object",
        "properties": {
            "device_id": {
                "type": "string",
                "description": "Unique device identifier",
            },
        },
        "required": ["device_id"],
    }
    output_schema = {
        "type": "object",
        "properties": {
            "device_id": {"type": "string"},
            "state": {"type": "object"},
            "online": {"type": "boolean"},
            "last_updated": {"type": "string"},
        },
    }

    def __init__(self, runtime: AgentRuntime):
        super().__init__()
        self.runtime = runtime

    async def execute(self, input_data: Dict[str, Any]) -> ToolResult:
        """Get device state from memory."""
        device_id = input_data["device_id"]

        device_data = await self.runtime.context.memory.get_device(device_id)

        if device_data:
            return ToolResult.ok(device_data)

        # If not in memory, try to query via MQTT if connected
        if self.runtime.mqtt and self.runtime.mqtt.connected:
            status_topic = self.runtime.mqtt.get_status_topic(device_id)
            # Publish a query message
            await self.runtime.mqtt.publish(
                f"{status_topic}/query",
                {"action": "get_state"}
            )
            # Return current memory state (may be empty)
            return ToolResult.ok(device_data or {
                "device_id": device_id,
                "state": {},
                "online": False,
            })

        return ToolResult.error(f"Device not found: {device_id}")


class DeviceStateSetTool(Tool):
    """Tool to set device state."""

    name = "device_state_set"
    description = "Set the state of a device (updates memory only)"
    input_schema = {
        "type": "object",
        "properties": {
            "device_id": {
                "type": "string",
                "description": "Unique device identifier",
            },
            "state": {
                "type": "object",
                "description": "Device state to set",
            },
            "online": {
                "type": "boolean",
                "description": "Device online status",
            },
            "device_name": {
                "type": "string",
                "description": "Human-readable device name",
            },
            "device_type": {
                "type": "string",
                "description": "Type of device (light, switch, etc.)",
            },
        },
        "required": ["device_id", "state"],
    }
    output_schema = {
        "type": "object",
        "properties": {
            "success": {"type": "boolean"},
            "device_id": {"type": "string"},
        },
    }

    def __init__(self, runtime: AgentRuntime):
        super().__init__()
        self.runtime = runtime

    async def execute(self, input_data: Dict[str, Any]) -> ToolResult:
        """Set device state in memory."""
        device_id = input_data["device_id"]

        # Get existing device data or create new
        device_data = await self.runtime.context.memory.get_device(device_id) or {}

        # Update fields
        device_data["device_id"] = device_id
        device_data["state"] = input_data.get("state", {})

        if "online" in input_data:
            device_data["online"] = input_data["online"]
        if "device_name" in input_data:
            device_data["device_name"] = input_data["device_name"]
            self.runtime.context.register_device_name(
                input_data["device_name"], device_id
            )
        if "device_type" in input_data:
            device_data["device_type"] = input_data["device_type"]

        # Save to memory
        await self.runtime.context.memory.set_device(device_id, device_data)

        return ToolResult.ok({
            "success": True,
            "device_id": device_id,
        })


class DeviceListTool(Tool):
    """Tool to list all devices."""

    name = "device_list"
    description = "List all known devices"
    input_schema = {
        "type": "object",
        "properties": {},
    }
    output_schema = {
        "type": "object",
        "properties": {
            "devices": {
                "type": "array",
                "items": {
                    "type": "object",
                    "properties": {
                        "device_id": {"type": "string"},
                        "device_name": {"type": "string"},
                        "device_type": {"type": "string"},
                        "online": {"type": "boolean"},
                    },
                },
            },
        },
    }

    def __init__(self, runtime: AgentRuntime):
        super().__init__()
        self.runtime = runtime

    async def execute(self, input_data: Dict[str, Any]) -> ToolResult:
        """List all devices from memory."""
        devices = await self.runtime.context.memory.list_devices()

        return ToolResult.ok({
            "devices": devices,
            "count": len(devices),
        })

File: pi-device-agent/src/tools/device-tools.ts
/**
 * Device tools for pi-mono
 *
 * Wraps Python Device Agent tools for pi-mono integration.
 */

import { Tool } from "pi-mono";
import { PythonBridge } from "../bridge";

/**
 * Create all device management tools
 */
export function createDeviceTools(bridge: PythonBridge): Tool[] {
  return [
    createDeviceControlTool(bridge),
    createDeviceStateGetTool(bridge),
    createDeviceStateSetTool(bridge),
    createDeviceListTool(bridge),
    createMQTTConnectTool(bridge),
    createMQTTPublishTool(bridge),
  ];
}

/**
 * Control a device (turn on/off, adjust settings)
 */
export function createDeviceControlTool(bridge: PythonBridge): Tool {
  return {
    name: "device_control",
    description: "Send a control command to an IoT device (turn on/off, set brightness, etc.)",
    inputSchema: {
      type: "object",
      properties: {
        device_id: {
          type: "string",
          description: "Unique device identifier (e.g., 'living_room_light')",
        },
        command: {
          type: "string",
          enum: ["on", "off", "toggle", "set"],
          description: "Command to execute",
        },
        params: {
          type: "object",
          description: "Additional parameters for 'set' command (e.g., { brightness: 80 })",
          default: {},
        },
      },
      required: ["device_id", "command"],
    },
    async execute(input) {
      return await bridge.callTool("device_control", input as Record<string, unknown>);
    },
  };
}

/**
 * Get device state
 */
export function createDeviceStateGetTool(bridge: PythonBridge): Tool {
  return {
    name: "device_state_get",
    description: "Get the current state of a device",
    inputSchema: {
      type: "object",
      properties: {
        device_id: {
          type: "string",
          description: "Unique device identifier",
        },
      },
      required: ["device_id"],
    },
    async execute(input) {
      return await bridge.callTool("device_state_get", input as Record<string, unknown>);
    },
  };
}

/**
 * Set device state (memory only)
 */
export function createDeviceStateSetTool(bridge: PythonBridge): Tool {
  return {
    name: "device_state_set",
    description: "Set device state in memory (does not send command to device)",
    inputSchema: {
      type: "object",
      properties: {
        device_id: {
          type: "string",
          description: "Unique device identifier",
        },
        state: {
          type: "object",
          description: "Device state object",
        },
        device_name: {
          type: "string",
          description: "Human-readable device name",
        },
        device_type: {
          type: "string",
          description: "Device type (light, switch, sensor, etc.)",
        },
        online: {
          type: "boolean",
          description: "Device online status",
        },
      },
      required: ["device_id", "state"],
    },
    async execute(input) {
      return await bridge.callTool("device_state_set", input as Record<string, unknown>);
    },
  };
}

/**
 * List all devices
 */
export function createDeviceListTool(bridge: PythonBridge): Tool {
  return {
    name: "device_list",
    description: "List all known devices",
    inputSchema: {
      type: "object",
      properties: {},
    },
    async execute() {
      return await bridge.callTool("device_list", {});
    },
  };
}

/**
 * Connect to MQTT broker
 */
export function createMQTTConnectTool(bridge: PythonBridge): Tool {
  return {
    name: "mqtt_connect",
    description: "Connect to an MQTT broker",
    inputSchema: {
      type: "object",
      properties: {
        broker: {
          type: "string",
          description: "MQTT broker hostname or IP address",
        },
        port: {
          type: "integer",
          description: "MQTT broker port",
          default: 1883,
        },
        username: {
          type: "string",
          description: "Username for authentication",
        },
        password: {
          type: "string",
          description: "Password for authentication",
        },
      },
      required: ["broker"],
    },
    async execute(input) {
      return await bridge.callTool("mqtt_connect", input as Record<string, unknown>);
    },
  };
}

/**
 * Publish MQTT message
 */
export function createMQTTPublishTool(bridge: PythonBridge): Tool {
  return {
    name: "mqtt_publish",
    description: "Publish a message to an MQTT topic",
    inputSchema: {
      type: "object",
      properties: {
        topic: {
          type: "string",
          description: "MQTT topic to publish to",
        },
        payload: {
          description: "Message payload (string or object)",
        },
        qos: {
          type: "integer",
          description: "Quality of Service level (0, 1, or 2)",
          default: 0,
        },
        retain: {
          type: "boolean",
          description: "Retain message",
          default: false,
        },
      },
      required: ["topic", "payload"],
    },
    async execute(input) {
      return await bridge.callTool("mqtt_publish", input as Record<string, unknown>);
    },
  };
}
